src/lib/account-number.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#extended-keys
+ * The max value that should be used for an hd key account number is 2^31 -1. In
+ * practice, we would never reach this via pure incremental account number
+ * generation, but we allow entering arbitrary account numbers in the UI, so we
+ * need to explicitly validate them. For whatever reason, the hdkey library
+ * we use throws for anything over 2^30.
+ */
+const MAX_UNSIGNED_INT_VALUE = 1073741824; // 2^30
+
+export function isValid32BitUnsignedInt(value: number) {
+  return (
+    Number.isInteger(value) && value >= 0 && value <= MAX_UNSIGNED_INT_VALUE
+  );
+}
+
+/**
+ * - The lowest possible sub-account number is 1 (0 is reserved for the "root" account).
+ * - If the set is empty, we just return 1.
+ * - Otherwise, we look for the highest number in the set and increment it by 1.
+ * - If the next incremented number is too big, we look back for the first gap in the numbers and use that instead.
+ * - If no gap is found (very unlikely), we throw an error.
+ */
+export function generateAccountNumber(accountNumbers: Set<number>): number {
+  if (accountNumbers.size === 0) {
+    return 1;
+  }
+
+  const sorted = Array.from(accountNumbers).sort((a, b) => a - b);
+  const currentHighestAccountNumber = sorted[sorted.length - 1];
+  const candidate = currentHighestAccountNumber + 1;
+
+  if (candidate <= MAX_UNSIGNED_INT_VALUE) {
+    return candidate;
+  }
+
+  // At most we look back 500 numbers. This is a bit arbitrary...  but the
+  // number of values could *technically* be 2^32 - 1, so we just limit the
+  // number of iterations to some reasonable value. The reason we look back for
+  // the highest available number instead of picking the lowest number is that
+  // the lowest number is more likely to have been used in the past and we're
+  // aiming to get a fresh wallet.
+  const maxLookBack = Math.max(sorted.length - 500, 0);
```

Comment:
How would we ever hit this? Looks like it can only happen if candidate = MAX_UNSIGNED_INT_VALUE, in which case maybe we should just return 0 or some default value.

Reply:
ahh, the way you can hit this is if you enter the max value directly into the recover thing. It's not likely, but possible. If that happens and you return just a default value, like zero, then you hit collisions with existing accounts, so we do need to do some search fora free spot. The question is just whether we search for the lowest or the highest value. I felt like searching for the highest value available makes the most sense bc it's far less likely that it would hit a recoverable account that you've already used before.